### ISSUE TITLE
Support for UWP and System XAML

### ISSUE THREAD

**Issue by asklar** [microsoft/CsWinRT#723]
Filing this here to track a dependency on cswinrt from https://github.com/microsoft/react-native-windows/issues/6539
We would like to be able to move away from C# UWP but we still need to target system xaml. So we would like the ability for cswinrt to target UWP apps and system XAML so that we can avoid a whole bunch of issues related to .net native and consuming nugets from C++ apps.

Related: https://github.com/microsoft/react-native-windows/issues/6539

**Comment by AdamBraden** [comment]
This would require custom type mapping for xaml to map onto inbox System XAML. Right now the custom type mappings go against WinUI.

**Comment by yoshiask** [comment]
I'm a little surprised that C#/WinRT doesn't already support system XAML. Just a noob here, but I had assumed that "lifting this projection support out of the compiler and runtime" would also entail decoupling from WinUI. Are there technical reasons why it can't/won't be done?

**Comment by driver1998** [comment]
Also this will unblock the XAML islands on .NET 5+ scenario, without this we'll have to wait for WinUI 3 XAML islands which is "scheduled to a later release".

**Comment by Jaiganeshkumaran** [comment]
I have a component that exposes a few platform XAML types. Before I could just use them from C# but now with C#/WinRT, it blacklists all those types from the projection, unlike every other WinRT projection. Even if you reserve certain mappings to Microsoft.UI.Xaml (eg: INotifyPropertyChanged), Windows.UI.Xaml types must be available even if you have Microsoft.UI.Xaml (from WinUI 3) referenced. C#/WinRT shouldn't be framework dependent.

<img src="https://cdn.discordapp.com/attachments/714838721790476289/925408560274341958/unknown.png">

**Comment by DontKnowOfTheGaps** [comment]
@angelazhangmsft  This issue is hanging on thin air and should be dropped right away.

UWP/CoreWindow was doomed to oblivion a loooong time ago. 
As .NET 5 / .NET 6 / .NET 7 and upwards versions themselves will _never_ support it , CsWinRT can't do very much, can it ?

**Comment by driver1998** [comment]
.NET 5 / 6 / 7 won't be compatible with AppContainer and UAP-only environments anytime soon, so UWP is not actually relevant here.
This is mainly for XAML islands.

**Comment by DontKnowOfTheGaaps** [comment]
so the title be modified to reflect XAML Island only. (Support for ~~UWP and~~ System XAML for XAML Islands.

`<RE>` well contrary to MS marketing, [AppContainer is not tied to any Window specific tech](https://github.com/microsoft/WindowsAppSDK/discussions/1071#discussioncomment-1012599) (CoreWindow aka UWP)`</RE>`

**Comment by dongle-the-gadget** [comment]
I don't think that it is necessary, you can actually create UWP apps with .NET 6, however the incompatibility with System XAML in stock C#/WinRT blocks that.

**Comment by driver1998** [comment]
That sounds like great news, I am always baffled by the lack of XAML islands support in .NET 5/6.

**Comment by driver1998** [comment]
@gus33000's fork of CsWinRT is at https://hub.fastgit.xyz/gus33000/CsWinRT/tree/wux.

**Comment by gus33000** [comment]
There are some roadblocks at the moment for having both Windows.UI.Xaml and Microsoft.UI.Xaml support in CsWinRT, which is why I haven't tried to do a PR yet for this. Mainly you have interop classes for .NET/WinRT that can currently only be applied to Windows.UI.Xaml or Microsoft.UI.Xaml. (As an example, CollectionChanged) It would be a bit messy at the moment to have support for both in the code. Maybe someone has an idea in order to make this a bit better?

**Comment by dongle-the-gadget** [comment]
> There are some roadblocks at the moment for having both Windows.UI.Xaml and Microsoft.UI.Xaml support in CsWinRT, which is why I haven't tried to do a PR yet for this. Mainly you have interop classes for .NET/WinRT that can currently only be applied to Windows.UI.Xaml or Microsoft.UI.Xaml. (As an example, CollectionChanged) It would be a bit messy at the moment to have support for both in the code. Maybe someone has an idea in order to make this a bit better?

Is there any specific reason why WinUI 3 needs C#/WinRT coupling? If no then I think the best approach is to just remove that coupling.

**Comment by Jaiganeshkumaran** [comment]
@gus33000 Most of the types eg: CollectionChanged are very similar across both WinUI 2.x and WinUI 3 so those can be projected as the same type while other types will be projected separately.

**Comment by lhak** [comment]
This should definitely be addressed because .net core 3.1 is going out of support by the end of the year and I doubt that winui 3 will be a good replacement for it by that time.

**Comment by dongle-the-gadget** [comment]
C#/WinRT 1.6.1 should already support building projections for system XAML, however building projections for libraries like the XamlApplication Toolkit or WinUI 2 causes the "class not registered" exception.

**Comment by driver1998** [comment]
> Mainly you have interop classes for .NET/WinRT that can currently only be applied to Windows.UI.Xaml or Microsoft.UI.Xaml
I guess it should be fine to split CsWinRT into two libraries right? CsWinRT for the core WinRT projection, and CsWinRT.WinUI for XAML related stuff (with separated assembly for system XAML and WinUI 3 XAML). 
Now with trimming working the worry of yet another DLL is not that relevant anymore.

**Comment by dongle-the-gadget** [comment]
> Maybe someone has an idea in order to make this a bit better?

How about adding a new property (i.e. `CsWinRTXamlNamespace`), like C++/WinRT is currently doing?

Of course, the property would be set to `Microsoft.UI.Xaml` by default.

**Comment by driver1998** [comment]
How much code in WinRT.Runtime is dependent on the XAML implementation? Can we generate entire classes like C++/WinRT does and eliminate the issue all together?

**Comment by dongle-the-gadget** [comment]
It powers some XAML types, as well as properly managing some error codes.

**Comment by Sergio0694** [comment]
If someone wants to try this out, #1421 adds System XAML support to CsWinRT, and Windows.UI.Xaml projections 🙂
